### ISSUE TITLE
Unexpected crash of LDPC decoder due to sparse find

### ISSUE THREAD

**Issue by pustovale** [NVlabs/sionna#186]
Few weeks ago I found that LDPC5GDecoder started crashing on my Windows machine 

```
Traceback (most recent call last):

  File c:\work\link_level\sionna_link_level\5g\main_5g_fec.py:170
    ber, bler = sim_ber(model,

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\sionna\utils\misc.py:633 in sim_ber
    outputs = mc_fun(batch_size=batch_size, ebno_db=ebno_dbs[i])

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\keras\utils\traceback_utils.py:70 in error_handler
    raise e.with_traceback(filtered_tb) from None

  File c:\work\link_level\sionna_link_level\5g\main_5g_fec.py:140 in call
    b_hat, tb_crc_status = self._tb_decoder(llr)

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\sionna\nr\tb_decoder.py:165 in call
    u_hat_cb = self._decoder(llr_cb)

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\sionna\fec\ldpc\decoding.py:1377 in call
    x_hat = super().call(llr_5g)

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\sionna\fec\ldpc\decoding.py:995 in call
    _, msg_vn, _ = tf.while_loop(dec_stop,

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\sionna\fec\ldpc\decoding.py:955 in dec_iter
    msg_vn = self._vn_update(msg_vn, llr_ch)

  File C:\Work\Software\Mamba\envs\sionna\lib\site-packages\sionna\fec\ldpc\decoding.py:518 in _vn_update
    x = tf.add(x, llr_ch)

InvalidArgumentError: Exception encountered when calling layer 'ldpc5g_decoder' (type LDPC5GDecoder).

{{function_node __wrapped__AddV2_device_/job:localhost/replica:0/task:0/device:CPU:0}} Incompatible shapes: [69320,100] vs. [15016,100] [Op:AddV2]

Call arguments received by layer 'ldpc5g_decoder' (type LDPC5GDecoder):
  • inputs=tf.Tensor(shape=(100, 1, 1, 14400), dtype=float32)
```

while the same code works on the Linux machine. After debugging I narrowed down the following code:

 ```
        # find all edges from variable and check node perspective
        self._cn_con, self._vn_con, _ = sp.sparse.find(pcm)

        # number of edges equals number of non-zero elements in the
        # parity-check matrix
        self._num_edges = len(self._vn_con)

        # permutation index to rearrange messages into check node perspective
        self._ind_cn = np.argsort(self._cn_con)

        # inverse permutation index to rearrange messages back into variable
        # node perspective
        self._ind_cn_inv = np.argsort(self._ind_cn)

        # generate row masks (array of integers defining the row split pos.)
        self._vn_row_splits = self._gen_node_mask_row(self._vn_con)
        self._cn_row_splits = self._gen_node_mask_row(
                                                    self._cn_con[self._ind_cn])
```

The code relies on the fact that `scipy.sparse.find` returns array self._vn_con sorted in descended order, while self._cn_con is unsorted. And this is indeed how it works on Linux machine. However, if we look at the documentation of `find`
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.find.html
there is no indication that sorting should be this way. In fact, on Windows they are sorted the opposite way:
self._cn_con is sorted in descending order while self._vn_con is returned unsorted. This causes the crash.

I tried to implement workaround
 ```
        # find all edges from variable and check node perspective
        self._cn_con, self._vn_con, _ = sp.sparse.find(pcm)

        #[Windows bug-fix -----------------------------------------------------
        temp=np.zeros(len(self._vn_con), dtype={'names':('vn_con', 'cn_con'),
                                'formats':('i4', 'i4')})
        temp['vn_con'] = self._vn_con
        temp['cn_con'] = self._cn_con

        idx = np.argsort(temp,order=['vn_con','cn_con'])
        self._vn_con = self._vn_con[idx]
        self._cn_con = self._cn_con[idx]
        #---------------------------------------------------------------------]

       # number of edges equals number of non-zero elements in the
       # parity-check matrix
       self._num_edges = len(self._vn_con)

       # permutation index to rearrange messages into check node perspective
       self._ind_cn = np.argsort(self._cn_con)

       # inverse permutation index to rearrange messages back into variable
       # node perspective
       self._ind_cn_inv = np.argsort(self._ind_cn)

       # generate row masks (array of integers defining the row split pos.)
       self._vn_row_splits = self._gen_node_mask_row(self._vn_con)
       self._cn_row_splits = self._gen_node_mask_row(
                                                   self._cn_con[self._ind_cn])
```

And it fixed the crash but the decoder still does not work properly on Windows. May be there is another unrelated issue.

**Comment by SebastianCa** [comment]
Hi,

thank you for reporting this issue. Unfortunately, I cannot reproduce the error using Windows 11. 
Could you please provide the following details:
- An example code snippet that reproduces the error (i.e., how you call the decoder)
- What is the remaining error message after your proposed fix?
- Your system configuration (Python, Scipy, Sionna and TF version)

A quick workaround to verify your theory would be to transpose the `pcm` before calling the ``sparse.find`` and swap `_cn_con` and `_vn_con` correspondingly.

**Comment by pustovale** [comment]
After applying my workaround there is no error messages any more, but the output of the code above is
BER=0.19,  BLER=1.0

Which does not make sense for low MCS and high SNR. On Linux BER and BLER are zeros.

**Comment by pustovale** [comment]
Transposing pcm and then swapping cn and vn gives the same result as my fix: no error messages, but incorrect decoding. 
Again, the later might be unrelated issue.

**Comment by pustovale** [comment]
Thanks. With scipy 1.10.1 it works

**Comment by pustovale** [comment]
@SebastianCa could you please publish the fix here or point out to the commit once it is ready. I would like to test whether it will help to fix this one
https://github.com/NVlabs/sionna/issues/190

**Comment by faycalaa** [comment]
Fixed by v0.15.1.
